### DISCUSSION TITLE
Daily Test Coverage Improver - Research and Plan

### DISCUSSION THREAD

**Discussion by github-actions** [kaito-project/aikit#674] [Ideas]
# Daily Test Coverage Improver - Research and Plan

## Executive Summary

**Current Test Coverage: 22.0%** (across 6 packages with tests)

AIKit is a Go-based project that builds container images with AI models and runtimes. The codebase has **only 6 test files** covering a small fraction of the 15 production Go files across 8 packages. Significant opportunities exist to improve coverage systematically.

---

### Repository Overview
- **Primary Language**: Go 1.24+
- **Build System**: Makefile + Docker buildx
- **Total Go Code**: ~2,400 lines across `pkg/` directory  
- **Test Files**: 6 files, ~914 lines of test code
- **Linting**: golangci-lint v2.1.6

### Package-by-Package Coverage

| Package | Coverage | Test File | Lines | Status |
|---------|----------|-----------|-------|--------|
| `pkg/aikit/config` | 90.0% | specs_test.go | 67 | ✅ Well tested |
| `pkg/build` | 23.5% | build_test.go | 393 | ⚠️ Validation only |
| `pkg/aikit2llb/inference` | 19.5% | 3 test files | 415 | ⚠️ Backend logic only |
| `pkg/utils` | 42.9% | util_test.go | 39 | ⚠️ Partial |
| `cmd/frontend` | 0.0% | none | 0 | ❌ No tests |
| `pkg/aikit2llb/finetune` | 0.0% | none | 0 | ❌ No tests |
| `pkg/version` | - | none | 0 | N/A (4 line const) |

### Major Coverage Gaps

**Critical (0% coverage):**
- `pkg/build/build.go` - Core build orchestration (539 lines)
  - `Build()`, `buildFineTune()`, `buildInference()`, `buildImage()`
  - `getAikitfileConfig()`, `getBuildArg()`, `parsePlatforms()`, `parseCacheOptions()`
  
- `pkg/build/args.go` - Build argument parsing (98 lines)
  - `parseBuildArgs()`, `generateInferenceConfig()`, `parseOCIURL()`
  
- `pkg/aikit2llb/inference/convert.go` - LLB conversion (173 lines)
  - `Aikit2LLB()`, `getBaseImage()`, `copyModels()`, `installCuda()`, `addLocalAI()`
  
- `pkg/aikit2llb/inference/download.go` - Model downloading (161 lines)
  - Protocol handlers: `handleOCI()`, `handleHTTP()`, `handleHuggingFace()`, `handleLocal()`
  
- `pkg/aikit2llb/inference/image.go` - Image configuration (54 lines)
- `pkg/aikit2llb/finetune/convert.go` - Fine-tuning (77 lines)
- `cmd/frontend/main.go` - CLI entry point (covered by integration tests)

**Moderate (< 50% coverage):**
- `pkg/utils/util.go` - `FileNameFromURL()` tested, but `Sh()`, `Shf()`, `Bashf()` untested

---

### Existing Commands (from Makefile)
```bash
# Unit tests with coverage
make test
# → go test -v ./... -race -coverprofile=coverage.txt -covermode=atomic

# Linting
make lint  
# → golangci-lint run -v ./... --timeout 5m

# Build validation (slow - minutes)
make build-test-model TEST_FILE=test/aikitfile-llama.yaml
make run-test-model
```

### Test Organization
- All tests follow Go convention: `*_test.go` alongside source
- Table-driven tests used extensively (good pattern)
- Test fixtures in `test/` directory (aikitfile-*.yaml configs)
- No mocking framework detected - using standard Go testing

### CI Workflows
- **lint.yaml**: Go setup → golangci-lint
- **test-docker-args.yaml**: Docker integration tests (3 protocols: OCI, HuggingFace, HTTPS)
- No dedicated unit test workflow found (tests likely run in combined workflows)

---

### Phase 1: High-Impact, Low-Complexity Areas (Weeks 1-2)
**Target: 35-40% total coverage**

1. **`pkg/build/args.go`** - Argument parsing functions
   - `parseBuildArgs()` - Test various build arg combinations
   - `generateInferenceConfig()` - Config generation logic  
   - `parseOCIURL()` - URL parsing edge cases
   - *Justification*: Pure functions, no external dependencies

2. **`pkg/utils/util.go`** - Shell utilities
   - `Sh()`, `Shf()`, `Bashf()` - Command execution helpers
   - *Justification*: Small file, easy to mock or stub

3. **`pkg/aikit2llb/inference/download.go`** - Protocol handlers
   - `extractModelName()` - String manipulation
   - `createCopyOptions()` - Option builders
   - *Justification*: Logic-heavy, testable without BuildKit

### Phase 2: Core Build Logic (Weeks 3-4)  
**Target: 50-55% total coverage**

4. **`pkg/build/build.go`** - Validation and defaulting functions
   - `parsePlatforms()` - Platform string parsing
   - `parseCacheOptions()` - Cache option parsing
   - `defaultsFineTune()` - Default value logic
   - *Note*: Skip `Build()`, `buildImage()` initially (require BuildKit context)

5. **`pkg/aikit2llb/inference/convert.go`** - LLB conversion helpers
   - `getBaseImage()` - Base image selection logic
   - Start testing `copyModels()` with mocked LLB state
   - *Justification*: Critical path, but requires careful mocking

### Phase 3: Integration & Complex Scenarios (Weeks 5+)
**Target: 60%+ total coverage**

6. **`pkg/aikit2llb/finetune/convert.go`** - Fine-tuning conversion
7. **Integration tests** for `pkg/build/build.go` core functions
8. **Edge cases** across all packages

---

### Patterns to Follow
```go
// Table-driven tests (existing pattern)
func TestFunction(t *testing.T) {
    tests := []struct {
        name    string
        input   Type
        want    Type
        wantErr bool
    }{
        // test cases
    }
    for _, tt := range tests {
        t.Run(tt.name, func(t *testing.T) {
            // test body
        })
    }
}
```

### Coverage Measurement
```bash
# Generate coverage
go test -v ./... -race -coverprofile=coverage.txt -covermode=atomic

# View summary
go tool cover -func=coverage.txt | grep total

# HTML report (for detailed analysis)
go tool cover -html=coverage.txt -o coverage.html
```

---

### 1. Snapshot Testing for LLB Graphs
The BuildKit LLB (low-level builder) generation is complex. Consider:
- Golden file tests for expected LLB graph outputs
- Serialize LLB state to JSON, compare against snapshots

### 2. Mock BuildKit Client
Major blocker for testing `pkg/build/build.go` and `pkg/aikit2llb/` is BuildKit dependency.
- Create lightweight mock implementing relevant buildkit interfaces
- Enables testing without Docker daemon

### 3. Property-Based Testing
For parsing functions (`parseOCIURL`, `parsePlatforms`, etc.):
- Use [gopter](https://github.com/leanovate/gopter) or similar
- Generate random inputs, verify invariants

### 4. Test Coverage Dashboard
- GitHub Actions artifact with coverage trends
- Compare coverage on PRs vs main branch
- Track per-package coverage metrics

---

## Questions for Maintainers

1. **BuildKit Testing**: Do you have preferences for mocking BuildKit clients in tests? Any existing patterns?

2. **Integration Test Scope**: Should I add tests that actually invoke BuildKit, or keep unit tests isolated?

3. **Test Fixtures**: The `test/` directory has many aikitfiles. Should I add minimal fixtures specifically for unit tests?

4. **Coverage Targets**: Is 60% overall coverage an acceptable long-term goal? Any packages that should be prioritized?

5. **Fine-Tuning Code**: `pkg/aikit2llb/finetune/` has 0% coverage. Is this actively maintained? Should it be prioritized?

6. **Shell Command Tests**: `pkg/utils/util.go` has shell execution functions. Mock or run actual commands in tests?

---

## Risks & Mitigations

| Risk | Impact | Mitigation |
|------|--------|-----------|
| BuildKit dependency blocks testing | High | Phase 1 focuses on BuildKit-free functions |
| Tests become brittle with mocking | Medium | Use table-driven tests, minimal mocking |
| Coverage metric gaming | Low | Focus on meaningful tests, not just coverage % |
| Slow test suite growth | Medium | Keep unit tests fast (<1s each), use `t.Parallel()` |

---

## Success Metrics

- ✅ **Overall coverage**: 22% → 60%+ (3x improvement)
- ✅ **Packages with tests**: 4 → 7 (cover all major packages)
- ✅ **Critical paths tested**: Build orchestration, LLB conversion, arg parsing
- ✅ **CI integration**: Coverage uploaded, tracked over time
- ✅ **Test suite speed**: Unit tests remain fast (<30s for all)

---

## Next Steps

Upon approval:
1. Create `.github/actions/daily-test-improver/coverage-steps/action.yml`
2. Validate coverage measurement workflow
3. Begin Phase 1 improvements (args.go, utils.go, download.go)
4. Report progress in this discussion weekly

**Ready to proceed!** Awaiting maintainer feedback on questions above.

> AI generated by [Daily Test Coverage Improver](https://github.com/kaito-project/aikit/actions/runs/18793291771)

**Comment by sozercan** [comment]
# Testing Questions - Answers

## 1. BuildKit Testing: Mocking Preferences

**No mocking is used.** The codebase has zero mock usage (no testify, gomock, etc.). Tests focus on pure function logic (config parsing, backend selection, validation). BuildKit LLB graph construction is not unit tested - only validated through integration tests in GitHub Actions workflows.

**Pattern**: Test input validation and business logic; skip BuildKit client interaction testing.

## 2. Integration Test Scope

**Keep unit tests isolated.** Integration tests already exist as separate GitHub Actions workflows (`test-docker.yaml`, `test-docker-gpu.yaml`, test-finetune.yaml). Unit tests don't invoke BuildKit.

**Don't add BuildKit invocation to unit tests** - would require Docker daemon and violate the fast loop (<1 min) principle.

## 3. Test Fixtures

**Use inline YAML strings for unit tests.** Current pattern in specs_test.go uses embedded YAML in test structs. The test directory aikitfiles are for integration tests only.

**Don't add dedicated unit test fixtures** - keep tests self-contained with inline data.

## 4. Coverage Targets

**Current: 22.0% overall**

Per-package:
- config: 90.0%
- inference: ~60-70%
- build: ~40-50%
- finetune: 0.0%
- utils: ~25%

**60% overall is acceptable** given that BuildKit orchestration code is hard to unit test and is covered by integration tests.

## 5. Fine-Tuning Code Priority

**Low priority.** finetune has 0% coverage but is actively maintained (has test-finetune.yaml workflow). Code is pure BuildKit LLB graph construction with complex GPU/NVIDIA driver requirements.

**Hard to unit test meaningfully** - already covered by integration tests on self-hosted GPU runners. Could add basic config validation tests only.

## 6. Shell Command Tests

**Don't mock or run shell commands.** Functions in util.go (`Sh`, `Shf`, `Bashf`) are simple wrappers around `llb.Args()` that return BuildKit run options - they don't execute shells during tests.

**Skip testing these** - they're implicitly validated by integration tests. Testing would only verify string formatting, which provides minimal value.

**Comment by github-actions** [comment]
## Phase 2 Complete: Coverage Configuration Added

I've successfully created and tested the coverage collection action. The configuration has been submitted as PR [Daily Test Coverage Improver - Updates to complete configuration](https://github.com/kaito-project/aikit/pulls).

### Initial Coverage Baseline

Successfully ran the coverage steps and confirmed:
- **Overall coverage: 22.0%** across all packages
- Coverage artifacts generated: `coverage.txt`, `coverage.html`, and `coverage-steps.log`
- All test steps execute cleanly with no errors

The action is ready for maintainer review. Once approved, I'll begin Phase 3 to systematically improve test coverage.

#### Related Items

- Pull Request: [#675](https://github.com/kaito-project/aikit/pull/675)


> AI generated by [Daily Test Coverage Improver](https://github.com/kaito-project/aikit/actions/runs/18797948162)
>
> To add this workflow in your repository, run `gh aw add githubnext/agentics/workflows/daily-test-improver.md@bf8d330e628f2c235cd9304310156bdbbfa61123`. See [usage guide](https://githubnext.github.io/gh-aw/tools/cli/).
